commons/all/src/main/java/org/infinispan/commons/netty/BranchlessParser.java

```diff
@@ -0,0 +1,207 @@
+      // A number using the whole space. We read the MSB from the last byte and process the remaining 3 bytes to
+      // handle the continuation bits.
+      if (readableBytes(buffer) <= 4 || (lastByte = buffer.getByte(buffer.readerIndex() + 4)) < 0) {
+         buffer.resetReaderIndex();
+         return 0;
+      }
+
+      buffer.skipBytes(5);
+      return lastByte << 28 | readInt(wholeOrMore);
+   }
+
+   private static int readInt(int continuation) {
+      // mix them up as per varint spec while dropping the continuation bits:
+      // 0x7F007F isolate the first byte and the third byte dropping the continuation bits
+      // 0x7F007F00 isolate the second byte and the fourth byte dropping the continuation bits
+      // the second and fourth byte are shifted to the right by 1, filling the gaps left by the first and third byte
+      // it means that the first and second bytes now occupy the first 14 bits (7 bits each)
+      // and the third and fourth bytes occupy the next 14 bits (7 bits each), with a gap between the 2s of 2 bytes
+      // and another gap of 2 bytes after the forth and third.
+      continuation = (continuation & 0x7F007F) | ((continuation & 0x7F007F00) >> 1);
+      // 0x3FFF isolate the first 14 bits i.e. the first and second bytes
+      // 0x3FFF0000 isolate the next 14 bits i.e. the third and forth bytes
+      // the third and forth bytes are shifted to the right by 2, filling the gaps left by the first and second bytes
+      return (continuation & 0x3FFF) | ((continuation & 0x3FFF0000) >> 2);
+   }
+
+   /**
+    * Reads the buffer for a long value encoded as varint.
+    *
+    * <p>
+    * <b>Warning:</b> check the buffer's reader index to verify if data was consumed.
+    * </p>
+    *
+    * @param buffer The buffer containing the encoded value.
+    * @return The encoded long, or <code>0</code>, if not enough bytes are available.
+    * @see #readRawVarint32(ByteBuf)
+    */
+   public static long readRawVarint64(ByteBuf buffer) {
+      if (readableBytes(buffer) <= Integer.BYTES || !hasCapacity(buffer, Long.BYTES))
+         return readRawVarint32(buffer);
+
+      long wholeOrMore = getLongLE(buffer);
+      long firstOneOnStop = ~wholeOrMore & 0x8080808080808080L;
+
+      // The value occupies all the bytes. We just unroll it and consume everything.
+      if (firstOneOnStop == 0) {
+         return readRawVarInt72(buffer, wholeOrMore);
+      }
+
+      // Consume the bytes containing the long.
+      int bitsToKeep = Long.numberOfTrailingZeros(firstOneOnStop) + 1;
+      buffer.skipBytes(bitsToKeep >> 3);
+
+      // Create a mask and create the continuation bytes for decoding.
+      long mask = firstOneOnStop ^ (firstOneOnStop - 1);
+      return readLong(wholeOrMore & mask);
+   }
+
+   private static long getLongLE(ByteBuf buffer) {
+      if (buffer instanceof ReplayableByteBuf rbb)
+         return getLongLE(rbb.internal());
+
+      return buffer.getLongLE(buffer.readerIndex());
+   }
```

Comment:
I don't like this :-1:.
This is necessary because the replaying buffer on the client side throws the REPLAY signal when we try to get the long value. We have already checked for the capacity, so this method should execute just fine.